### PR TITLE
image store local and s3

### PR DESCRIPTION
**Pull request by lospro7**

### CODE REVIEW

File: imagestore/factory.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+func (this *Factory) NewS3() *S3ImageStore {
+	return &S3ImageStore{
+		bucketName: "gophergala",
+		storeRoot:  "original",
+	}
+}
+
+func (this *Factory) NewLocal() *LocalImageStore {
+	return &LocalImageStore{
+		storeRoot: "/Users/jarvis/imagestore",
```
Comment (briankassouf) [inline imagestore/factory.go]: this should be moved into the config file

File: imagestore/factory.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package imagestore
+
+type Factory struct {
+	config string
+}
+
+func (this *Factory) NewS3() *S3ImageStore {
+	return &S3ImageStore{
+		bucketName: "gophergala",
```
Comment (jacobgreenleaf) [inline imagestore/factory.go]: config

File: imagestore/localstore.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+func (this *LocalImageStore) Exists(obj *StoreObject) bool {
+	if _, err := os.Stat(this.storeRoot + "/" + obj.Path); os.IsNotExist(err) {
+		return false
+	}
+
+	return true
+}
+
+func (this *LocalImageStore) Save(src string, obj *StoreObject) error {
+	// open input file
```
Comment (jacobgreenleaf) [inline imagestore/localstore.go]: change `src string` to `fi io.File`?

File: imagestore/factory.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+func (this *Factory) NewLocal() *LocalImageStore {
+	return &LocalImageStore{
+		storeRoot: "/Users/jarvis/imagestore",
+	}
+}
+
+func (this *Factory) NewStoreObject(name string, mime string) *StoreObject {
+	return &StoreObject{
+		Path:     name,
```
Comment (jacobgreenleaf) [inline imagestore/factory.go]: should this be called `name` or `path`? What if we had an `InMemoryStore`?

File: imagestore/s3store.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	return true
+}
+
+func (this *S3ImageStore) Save(src string, obj *StoreObject) error {
+	auth, err := aws.EnvAuth()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	client := s3.New(auth, aws.USEast)
```
Comment (Jasdev) [inline imagestore/s3store.go]: Add ability to customize region

File: imagestore/s3store.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+func NewS3ImageStore(bucket string, root string) *S3ImageStore {
+	return &S3ImageStore{
+		bucketName: bucket,
+		storeRoot:  root,
+	}
+}
+
+func (this *S3ImageStore) Exists(obj *StoreObject) bool {
+	return true
```
Comment (jacobgreenleaf) [inline imagestore/s3store.go]: panic not implemented

File: imagestore/factory.go (hunk 1 above)
Comment (lospro7) [inline imagestore/factory.go]: it will be

File: imagestore/factory.go (hunk 4 above)
Comment (lospro7) [inline imagestore/factory.go]: in that case it should probably be a name

File: imagestore/s3store.go (hunk 5 above)
Comment (jacobgreenleaf) [inline imagestore/s3store.go]: why does it make a new `s3` per call to `Save()`?

Comment (lospro7) [inline imagestore/s3store.go]: I don't it has to

**Comment by briankassouf** [comment]
:+1:

File: imagestore/localstore.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+func (this *LocalImageStore) Exists(obj *StoreObject) (bool, error) {
+	if _, err := os.Stat(this.toPath(obj)); os.IsNotExist(err) {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (this *LocalImageStore) Save(src string, obj *StoreObject) error {
```
Comment (jacobgreenleaf) [inline imagestore/localstore.go]: should this be `source io.Reader`?

File: imagestore/s3store.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+	return &S3ImageStore{
+		bucketName: bucket,
+		storeRoot:  root,
+		region:     region,
+		client:     client,
+	}
+}
+
+func (this *S3ImageStore) Exists(obj *StoreObject) (bool, error) {
+	bucket := this.client.Bucket(this.bucketName)
```
Comment (jacobgreenleaf) [inline imagestore/s3store.go]: put `bucket` in instance field?

Comment (lospro7) [inline imagestore/s3store.go]: In store object?

**Pull request merged**